File: CODE/Sln/Backup/Model/SaleManage/AskPrice/AskPrice.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace OA.Model
{
    /// <summary>
    /// 询价单
    /// </summary>
    public class AskPrice
    {
        private string _apid;
        private string _apcode;
        private string _askdate;
        private string _clientname;
        private string _clientcontact;
        private string _clienttel;
        private string _clientaddress;
        private string _materialcode;
        private string _materialname;
        private string _specs;
        private string _routing;
        private decimal _planprice;
        private string _issued;
        private int _issuedcount;
        private decimal _actualprice;
        private string _paytypeid;
        private string _trackdescription;
        private string _clientsurvey;
        private string _apremark;
        private string _creator;
        private DateTime _createTime;
        private string _editor;
        private DateTime _edittime;
        private string _firstchecker;
        private DateTime _firstchecktime;
        private string _firstcheckview;
        private string _state;
        private string _secondcheckername;
        private string _readername;
        private decimal _qty;
        private string _unitid;
        private string _istax;
        private string _isshipping;
        private List<APSecondCheck> _sclist;
        private List<APReader> _rlist;

        /// <summary>
        /// 主键
        /// </summary>
        public string APID
        {
            set { _apid = value; }
            get { return _apid; }
        }
        /// <summary>
        /// 询价单编号
        /// </summary>
        public string APCode
        {
            set { _apcode = value; }
            get { return _apcode; }
        }
        /// <summary>
        /// 询价日期
        /// </summary>
        public string AskDate
        {
            set { _askdate = value; }
            get { return _askdate; }
        }
        /// <summary>
        /// 客户名称
        /// </summary>
        public string ClientName
        {
            set { _clientname = value; }
            get { return _clientname; }
        }
        /// <summary>
        /// 客户联系人
        /// </summary>
        public string ClientContact
        {
            set { _clientcontact = value; }
            get { return _clientcontact; }
        }
        /// <summary>
        /// 客户电话
        /// </summary>
        public string ClientTel
        {
            set { _clienttel = value; }
            get { return _clienttel; }
        }
        /// <summary>
        /// 客户地址
        /// </summary>
        public string ClientAddress
        {
            set { _clientaddress = value; }
            get { return _clientaddress; }
        }
        /// <summary>
        /// 产品编号
        /// </summary>
        public string MaterialCode
        {
            set { _materialcode = value; }
            get { return _materialcode; }
        }
        /// <summary>
        /// 产品名称
        /// </summary>
        public string MaterialName
        {
            set { _materialname = value; }
            get { return _materialname; }
        }
        /// <summary>
        /// 规格型号
        /// </summary>
        public string Specs
        {
            set { _specs = value; }
            get { return _specs; }
        }
        /// <summary>
        /// 工艺
        /// </summary>
        public string Routing
        {
            set { _routing = value; }
            get { return _routing; }
        }
        /// <summary>
        /// 价格
        /// </summary>
        public decimal PlanPrice
        {
            set { _planprice = value; }
            get { return _planprice; }
        }
        /// <summary>
        /// 有无下达单据
        /// 0：没下达
        /// 1：已下达
        /// </summary>
        public string Issued
        {
            set { _issued = value; }
            get { return _issued; }
        }
        /// <summary>
        /// 下达数量
        /// </summary>
        public int IssuedCount
        {
            set { _issuedcount = value; }
            get { return _issuedcount; }
        }
        /// <summary>
        /// 金额
        /// </summary>
        public decimal ActualPrice
        {
            set { _actualprice = value; }
            get { return _actualprice; }
        }
        /// <summary>
        /// 付款方式ID
        /// </summary>
        public string PayTypeID
        {
            set { _paytypeid = value; }
            get { return _paytypeid; }
        }
        /// <summary>
        /// 跟踪情况
        /// </summary>
        public string TrackDescription
        {
            set { _trackdescription = value; }
            get { return _trackdescription; }
        }
        /// <summary>
        /// 客户调查
        /// </summary>
        public string ClientSurvey
        {
            set { _clientsurvey = value; }
            get { return _clientsurvey; }
        }
        /// <summary>
        /// 备注
        /// </summary>
        public string APRemark
        {
            set { _apremark = value; }
            get { return _apremark; }
        }
        /// <summary>
        /// 创建人
        /// </summary>
        public string Creator
        {
            set { _creator = value; }
            get { return _creator; }
        }
        /// <summary>
        /// 创建时间
        /// </summary>
        public DateTime CreateTime
        {
            set { _createTime = value; }
            get { return _createTime; }
        }
		/// <summary>
        /// 修改人ID
		/// </summary>
		public string Editor
		{
			set{ _editor=value;}
			get{return _editor;}
		}
		/// <summary>
        /// 修改时间
		/// </summary>
		public DateTime EditTime
		{
			set{ _edittime=value;}
			get{return _edittime;}
		}
		/// <summary>
        /// 初审人ID
		/// </summary>
		public string FirstChecker
		{
			set{ _firstchecker=value;}
			get{return _firstchecker;}
		}
		/// <summary>
        /// 初审时间
		/// </summary>
		public DateTime FirstCheckTime
		{
			set{ _firstchecktime=value;}
			get{return _firstchecktime;}
		}
		/// <summary>
        /// 初审意见
		/// </summary>
		public string FirstCheckView
		{
			set{ _firstcheckview=value;}
			get{return _firstcheckview;}
		}
		/// <summary>
        /// 单据状态
		/// 1:编制
        /// 2:提交
        /// 3:审核通过
        /// 4:审核不通过
        /// 5:批准通过
        /// 6:批准不通过
        /// 7:关闭
		/// </summary>
		public string State
		{
			set{ _state=value;}
			get{return _state;}
		}
		/// <summary>
        /// 复核人姓名
		/// </summary>
		public string SecondCheckerName
		{
			set{ _secondcheckername=value;}
			get{return _secondcheckername;}
		}
		/// <summary>
        /// 分阅人姓名
		/// </summary>
		public string ReaderName
		{
			set{ _readername=value;}
			get{return _readername;}
		}
        /// <summary>
        /// 复审人
        /// </summary>
        public List<APSecondCheck> SCList
        {
            set { _sclist = value; }
            get { return _sclist; }
        }
        /// <summary>
        /// 分阅人
        /// </summary>
        public List<APReader> RList
        {
            set { _rlist = value; }
            get { return _rlist; }
        }
        /// <summary>
        /// 数量
        /// </summary>
        public decimal Qty
        {
            set { _qty = value; }
            get { return _qty; }
        }
        /// <summary>
        /// 计量单位
        /// </summary>
        public string UnitID
        {
            set { _unitid = value; }
            get { return _unitid; }
        }
        /// <summary>
        /// 是否含税
        /// </summary>
        public string IsTax
        {
            set { _istax = value; }
            get { return _istax; }
        }
        /// <summary>
        /// 是否含运费
        /// </summary>
        public string IsShipping
        {
            set { _isshipping = value; }
            get { return _isshipping; }
        }
    }
}

File: CODE/Src/PlatForm/OA/SalesManage/AskPrice/AskPrice.js
//询价单信息格式化对象
var apFormatter = {
    //询价单状态
    apState: {
        src: [{ value: '1', text: '编制' }, { value: '2', text: '提交初审' }, { value: '3', text: '初审通过' }, { value: '4', text: '初审不通过' }, { value: '5', text: '提交复审' }, { value: '6', text: '复审通过' }, { value: '7', text: '复审不通过' }, { value: '8', text: '关闭' }],
        format: function (value) {
            if (gFunc.isNull(value)) {
                return "";
            }
            var rst = "";
            for (var idx = 0; idx < this.src.length; idx++) {
                if (value == this.src[idx].value) {
                    rst = this.src[idx].text;
                    break;
                }
            }
            return rst;
        }
    },
    //询价单类别
    apType: {
        src: [{ value: 'OffsetPrint', text: '胶印新产品' }, { value: 'SilkScreen', text: '丝印新产品' }],
        format: function (value) {
            if (gFunc.isNull(value)) {
                return "";
            }
            var rst = "";
            for (var idx = 0; idx < this.src.length; idx++) {
                if (value == this.src[idx].value) {
                    rst = this.src[idx].text;
                    break;
                }
            }
            return rst;
        }
    },
};

//询价单列表对象
var askprice = {
    stateConf: {
        add: '0',
        edit: '1',
        view: '2'
    },
    grid: $('#grid'),
    gridSOItem: $('#gridAPItem'),
    formSearch: $('#searchForm'),
    btnSearch: $('#btnSearch'),
    txtSearchAPCode: $('#txtSearchAPCode'),
    cbAPType: $('#cbAPType'),
    dateSearchAskDateBegin: $('#dateSearchAskDateBegin'),
    dateSearchAskDateEnd: $('#dateSearchAskDateEnd'),
    cardFormWidth: 700,
    cardFormHeight: 600,
    approvalFormWidth: 400,
    approvalFormHeight: 240,
    cardFormUrl: 'AskPriceAdd.aspx',
    searchUrl: 'AskPriceService.asmx/GetList',
    init: function (funcType) {
        askprice.initgrid();
        askprice.bindingEvents();
        askprice.formSearch.children('div').css({ 'float': 'left', 'padding-left': '8px' });
        askprice.cbAPType.combobox({
            data: apFormatter.apType.src,
            valueField: 'value',
            textField: 'text'
        });
    },
    //绑定（注册）事件
    bindingEvents: function () {
        askprice.btnSearch.click(askprice.doSearch);
    },
    initgrid: function () {
        gFunc.initGridPublic(askprice.grid, {
            title: '询价单列表',
            icon: 'icon-edit',
            key: 'APID',
            url: askprice.searchUrl,
            toolbar: [{
                id: 'btnAdd',
                text: '新增',
                iconCls: 'icon-add',
                handler: askprice.addAskPrice
            }, "-", {
                id: 'btnView',
                text: '查看',
                iconCls: 'icon-search',
                handler: askprice.viewAskPrice
            }, "-", {
                id: 'btnEdit',
                text: '修改',
                iconCls: 'icon-edit',
                handler: askprice.editAskPrice
            }, "-", {
                id: 'btnDelete',
                text: '删除',
                iconCls: 'icon-remove',
                handler: askprice.deleteRowBatch
            }, "-", {
                id: 'btnSubmitToFirstCheck',
                text: '提交初审',
                iconCls: 'icon-search',
                handler: askprice.submitToFirstCheck
            }, "-", {
                id: 'btnSubmitToSecondCheck',
                text: '提交复审',
                iconCls: 'icon-edit',
                handler: askprice.submitToSecondCheck
            }, "-", {
                id: 'btnSubmitToReader',
                text: '设置分阅人',
                iconCls: 'icon-remove',
                handler: askprice.submitToRead
            }],
            columns: [[
                { field: 'ck', title: '', width: 100, align: 'center', checkbox: true },
                { field: 'APID' },
                { field: 'APCode', title: '询价单编号', width: 100, align: 'center' },
                {
                    field: 'State', title: '询价单状态', width: 80, align: 'center',
                    formatter: function (value, row, index) { return apFormatter.apState.format(value); }
                },
                {
                    field: 'APType', title: '询价单类型', width: 80, align: 'center',
                    formatter: function (value, row, index) { return apFormatter.apType.format(value); }
                },
                { field: 'AskDate', title: '询价日期', align: 'center', width: 80, formatter: formatHandler.date.format },
                { field: 'ClientID' },
                { field: 'Client_Name', title: '客户名称', align: 'center', width: 100 },
                { field: 'Client_Contact', title: '客户联系人', align: 'center', width: 100 },
                { field: 'Client_Tel', title: '客户电话', align: 'center', width: 100 },
                { field: 'Client_Address', title: '客户地址', align: 'center', width: 100 },
                { field: 'PayTypeID' },
                { field: 'PayType_Name', title: '付款方式', align: 'center', width: 80 },
                { field: 'TrackDescription', title: '跟踪情况', align: 'center', width: 120 },
                { field: 'ClientSurvey', title: '客户调查', align: 'center', width: 120 },
                { field: 'APRemark', title: '备注', align: 'center', width: 100 },
                { field: 'Creator' },
                { field: 'Creator_Name', title: '创建人', align: 'center', width: 80 },
                { field: 'CreateTime', title: '创建时间', align: 'center', width: 130 },
                { field: 'Editor' },
                { field: 'Editor_Name', title: '修改人', align: 'center', width: 80 },
                { field: 'EditTime', title: '修改时间', align: 'center', width: 130 },
                { field: 'FirstChecker' },
                { field: 'FirstChecker_Name', title: '初审人', align: 'center', width: 80 },
                { field: 'FirstCheckTime', title: '初审时间', align: 'center', width: 130 },
                { field: 'FirstCheckView', title: '初审意见', align: 'center', width: 100 },
                { field: 'SecondCheckerName', title: '复审人', align: 'center', width: 80 },
                { field: 'ReaderName', title: '分阅人', align: 'center', width: 60 }
            ]],
            hidecols: ['APID', 'ClientID', 'PayTypeID', 'Creator', 'Editor', 'FirstChecker'],
            singleSelect: false
        });
    },
    //编辑
    addAskPrice: function () {
        location.href = askprice.cardFormUrl + '?state=0';
    },
    viewAskPrice: function () {
        var checkedRows = askprice.grid.datagrid('getChecked');
        if (gFunc.isNull(checkedRows) || checkedRows.length < 1) {
            $.messager.alert('提示', '请选择要查看的数据');
            return;
        }
        if (checkedRows.length > 1) {
            $.messager.alert('提示', '请选择一条数据查看');
            return;
        }
        //location.href = askprice.cardFormUrl + '?' + encodeURI('state=2&apdata=' + JSON.stringify(checkedRows[0]));//
        location.href = askprice.cardFormUrl + '?' + encodeURI('state=2&apId=' + checkedRows[0].APID);//
    },
    editAskPrice: function () {
        var rows = askprice.grid.datagrid('getChecked');
        ////console.log(rows.length);
        if (gFunc.isNull(rows) || rows.length < 1) {
            $.messager.alert('提示', '请选择要修改的数据');
            return;
        }
        if (rows.length > 1) {
            $.messager.alert('提示', '只能修改一条数据');
            return;
        }
        //
        if (rows[0].State !== '1' && rows[0].State !== '4' && rows[0].State !== '7') {
            $.messager.alert('提示', '请选择编制、初审不通过或复审不通过状态的询价单');
            return;
        }
        //location.href = askprice.cardFormUrl + '?' + encodeURI('state=1&apdata=' + JSON.stringify(rows[0]));//
        location.href = askprice.cardFormUrl + '?' + encodeURI('state=1&apId=' + rows[0].APID);//
    },
    deleteRowBatch: function () {
        var delCheckedRows = askprice.grid.datagrid('getChecked');
        if (gFunc.isNull(delCheckedRows) || delCheckedRows.length < 1) {
            $.messager.alert('提示', '请选择要删除的数据');
            return;
        }
        var illegalRows = $.grep(delCheckedRows, function (row, idx) {
            return row.State !== '1' && row.State !== '4'
                && row.State !== '7' && row.State !== '8';//过滤条件：非(编制、初审不通过、复审不通过、关闭)
        });
        if (illegalRows.length > 0) {
            $.messager.alert('提示', '请选择编制、初审不通过、复审不通过或关闭状态的询价单');
            return;
        }

        $.messager.confirm('询问', '确定要删除所选数据吗？', function (result) {
            if (result) {
                //2、删除选中行中以保存的部分（这部分提交到服务端删除，然后刷新列表）
                var ids = [];
                $.each(delCheckedRows, function (index, row) {
                    ids.push(row.APID);
                });
                //这里json序列化的目标一定是一个数组，否则，后台解析（解析为列表）时会出错
                $.post('AskPriceService.asmx/Delete', JSON.stringify(ids), function (result) {
                    if (result && result.code) {
                        //重新加载
                        askprice.grid.datagrid('reload');
                    }
                });
            }
        });
    },
    doSearch: function () {
        //收集查询条件
        var searchParams = askprice.formSearch.serializeToJson(true);
        console.log(JSON.stringify(searchParams));
        //重新查询
        askprice.grid.datagrid("reload", searchParams);
    },
    //提交
    submitToFirstCheck: function () {
        //获取选中行
        var checkedRows = askprice.grid.datagrid('getChecked');
        if (gFunc.isNull(checkedRows) || checkedRows.length < 1) {
            $.messager.alert('提示', '请选择数据');
            return;
        }
        var illegalRows = $.grep(checkedRows, function (row, idx) {
            return row.State !== '1' && row.State !== '4';
        });
        if (illegalRows.length > 0) {
            $.messager.alert('提示', '请选择编制或初审不通过状态的询价单');
            return;
        }
        //提交初审
        showPopGridHelp(400, 300, true, helpInitializer.user, askprice.helpReceiver.submitToFirstChecker, null);
    },
    submitToSecondCheck: function () {
        //获取选中行
        var checkedRows = askprice.grid.datagrid('getChecked');
        if (gFunc.isNull(checkedRows) || checkedRows.length < 1) {
            $.messager.alert('提示', '请选择数据');
            return;
        }
        var illegalRows = $.grep(checkedRows, function (row, idx) {
            return row.State !== '3' && row.State !== '7';
        });
        if (illegalRows.length > 0) {
            $.messager.alert('提示', '请选择初审通过或复审不通过状态的询价单');
            return;
        }
        //提交初审
        showPopGridHelp(400, 300, true, helpInitializer.user, askprice.helpReceiver.submitToSecondChecker, null);
    },
    submitToRead: function () {
        //获取选中行
        var checkedRows = askprice.grid.datagrid('getChecked');
        if (gFunc.isNull(checkedRows) || checkedRows.length < 1) {
            $.messager.alert('提示', '请选择数据');
            return;
        }
        //提交分阅
        showPopGridHelp(400, 300, true, helpInitializer.user, askprice.helpReceiver.submitToReader, null);
    },
    helpReceiver: {
        submitToFirstChecker: function (userData) {
            if (gFunc.isNull(userData) || gFunc.isNull(userData.UserID)) {
                $.messager.alert('提示', '请选择初审人');
                return;
            }
            //获取选中行
            var checkedRows = askprice.grid.datagrid('getChecked');
            if (gFunc.isNull(checkedRows) || checkedRows.length < 1) {
                $.messager.alert('提示', '请选择数据');
                return;
            }
            var apIds = [];
            $.each(checkedRows, function (index, row) {
                apIds.push(row.APID);
            });
            var ajaxResult = false;
            $.ajax({
                type: 'post',
                url: 'AskPriceService.asmx/SubmitToFirstChecker',
                data: 'userId=' + userData.UserID + '&apIds=' + JSON.stringify(apIds),
                async: false,//同步请求
                success: function (result) {
                    if (result && result.code) {
                        ajaxResult = true;
                        askprice.grid.datagrid('reload');
                        ////console.log('askprice,ajax succeed');
                    } else {
                        ////console.log('askprice,ajax fail');
                        ajaxResult = false;
                    }
                },
                error: function () {
                    //console.log('askprice,ajax error');
                    ajaxResult = false;
                }
            });
        },
        submitToSecondChecker: function (userData) {
            if (gFunc.isNull(userData) || gFunc.isNull(userData.UserID)) {
                $.messager.alert('提示', '请选择复审人');
                return;
            }
            //获取选中行
            var checkedRows = askprice.grid.datagrid('getChecked');
            if (gFunc.isNull(checkedRows) || checkedRows.length < 1) {
                $.messager.alert('提示', '请选择数据');
                return;
            }
            var apIds = [];
            $.each(checkedRows, function (index, row) {
                apIds.push(row.APID);
            });
            var ajaxResult = false;
            $.ajax({
                type: 'post',
                url: 'AskPriceService.asmx/SubmitToSecondChecker',
                data: 'userId=' + userData.UserID + '&apIds=' + JSON.stringify(apIds),
                async: false,//同步请求
                success: function (result) {
                    if (result && result.code) {
                        ajaxResult = true;
                        askprice.grid.datagrid('reload');
                        //console.log('askprice,ajax succeed');
                    } else {
                        //console.log('askprice,ajax fail');
                        ajaxResult = false;
                    }
                },
                error: function () {
                    //console.log('askprice,ajax error');
                    ajaxResult = false;
                }
            });
        },
        submitToReader: function (userData) {
            if (gFunc.isNull(userData) || gFunc.isNull(userData.UserID)) {
                $.messager.alert('提示', '请选择分阅人');
                return;
            }
            //获取选中行
            var checkedRows = askprice.grid.datagrid('getChecked');
            if (gFunc.isNull(checkedRows) || checkedRows.length < 1) {
                $.messager.alert('提示', '请选择数据');
                return;
            }
            var apIds = [];
            $.each(checkedRows, function (index, row) {
                apIds.push(row.APID);
            });
            var ajaxResult = false;
            $.ajax({
                type: 'post',
                url: 'AskPriceService.asmx/SubmitToReader',
                data: 'userId=' + userData.UserID + '&apIds=' + JSON.stringify(apIds),
                async: false,//同步请求
                success: function (result) {
                    if (result && result.code) {
                        ajaxResult = true;
                        askprice.grid.datagrid('reload');
                        //console.log('askprice,ajax succeed');
                    } else {
                        //console.log('askprice,ajax fail');
                        ajaxResult = false;
                    }
                },
                error: function () {
                    //console.log('askprice,ajax error');
                    ajaxResult = false;
                }
            });
        },
    }
}